### PR TITLE
Creates EmploymentDetail model

### PR DESCRIPTION
**Pull request by mohinid**
## Notion card
https://www.notion.so/saeloun/Add-tables-for-personal-employment-Allocated-devices-compensation-71ef49d0c38e4242bd2998120141372d

## Summary
1. Creates employment_details table with appropriate columns.
2. Creates 1-1 association between CompanyUser and EmploymentDetail model.
3. Adds validation for the EmploymentDetail model.
4. Add RSpec for both EmploymentDetail and CompanyUser models.

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Added validations & RSpecs for User and EmploymentDetail models.

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [x] I have added automated tests for my code

### CODE REVIEW

File: spec/models/employment_detail_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+RSpec.describe EmploymentDetail, type: :model do
+  subject(:employment_detail) { build(:employment_detail) }
+
+  describe "Validations" do
+    describe "Associations" do
+      it { is_expected.to belong_to(:user) }
+    end
+
+    describe "validate presence of" do
```
Comment (rohitjoshixyz) [inline spec/models/employment_detail_spec.rb:13]:
```suggestion
    describe "Presence" do
```

File: spec/models/employment_detail_spec.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      it { is_expected.to validate_presence_of(:employment_type) }
+      it { is_expected.to validate_presence_of(:joined_at) }
+      it { is_expected.to validate_presence_of(:employee_id) }
+    end
+
+    describe "validate comparisons" do
+      it "resignation date should be later than joining date" do
+        expect(employment_detail.resigned_at).to be > employment_detail.joined_at
+      end
+    end
```
Comment (rohitjoshixyz) [inline spec/models/employment_detail_spec.rb:24]: What happens when either of them is `nil`? Add a spec for each

File: spec/factories/employment_details.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :employment_detail do
+    user
+    designation { "SDE" }
+    employment_type { "Salaried" }
+    joined_at { Faker::Date.between(from: "2020-01-01", to: "2021-01-01") }
+    resigned_at { Date.today }
```
Comment (rohitjoshixyz) [inline spec/factories/employment_details.rb:9]: This factory will always create a resigned employee by default.
Let's keep `resigned_at` nil by default

**Review by rohitjoshixyz — COMMENTED** [review]
@mohinid Added a few suggestions

File: spec/models/employment_detail_spec.rb (hunk 2 above)
Comment (mohinid) [reply spec/models/employment_detail_spec.rb:24]: Joined_at cannot be nill. I have validation added for the same. Will add test for nil resigned_at date.

File: app/models/user.rb [hunk 4]
```diff
@@ -60,6 +60,7 @@ class User < ApplicationRecord
   has_many :identities, dependent: :delete_all
   has_one_attached :avatar
   rolify strict: true
+  has_one :employment_detail, dependent: :destroy
```
Comment (akhilgkrishnan) [inline app/models/user.rb:63]: can you move this on line:62

**Review by akhilgkrishnan — COMMENTED** [review]
@mohinid I have got a couple of doubts related to this flow because, As per the Miru workflow the User can be part of multiple companies and the employee details vary with the position handled in each company. So we should add a scope for the company model.

File: app/models/user.rb (hunk 4 above)
Comment (mohinid) [reply app/models/user.rb:63]: done

**Comment by mohinid** [comment]
> @mohinid I have got a couple of doubts related to this flow because, As per the Miru workflow the User can be part of multiple companies and the employee details vary with the position handled in each company. This EmploymentDetails model hasn't given an idea about the company which the employee is associated with. So we should add a scope for the company model.

Hi @akhilgkrishnan , let's discuss today.

**Comment by rohitjoshixyz** [comment]
@akhilgkrishnan Good catch: we could associate the `employment_details` table with the `company_users` table instead of the `user` model, which should address this concern @mohinid

**Comment by mohinid** [comment]
Removed User reference and added reference of company_detail with employment_detail. @akhilgkrishnan  @rohitjoshixyz 
All tests passed. Please review.

Screenshot 2022-06-13 at 3 14 13 PM

**Review by rohitjoshixyz — APPROVED** [review]
LGTM 🎉

**Review by akhilgkrishnan — APPROVED** [review]
LGTM. Thanks @mohinid for making the requested changes
